### PR TITLE
Frontend documentation

### PR DESCRIPTION
**Pull request by floriank**
This will add desperately needed documentation for frontend-related topics for OpenProject.

The main concern here is to:
- document the processes involved with building the frontend
- convey information on trap doors and dark corners of the frontend code
- document any quirks and tools that are used throughout the process.

Initially, this will cover the AngularJS parts of the frontend. Ideally however, this will also cover some of the legacy codebase found in the dark reaches of places like `application.js.erb`.

Please look out for interesting grammar and typos - I am not the best playwright in the world, despite contrary information.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://floriank-feature-frontend-documentation.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://floriank-feature-frontend-documentation.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by floriank** [comment]
@ulferts @NobodysNightmare can someone please label this as work in progress?

**Comment by floriank** [comment]
I am mostly through with the heavier lifting for `angular`. There is a lot of shady stuff to document in the Rails legacy  views and the dark places like `application.js.erb`.

In the meantime, if you have questions regarding the frontend development, I'd like to put an FAQ into the documentation.

File: frontend/doc/API.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+- [APIv3](http://opf.github.io/apiv3-doc/)
+- APIv2 docs are located in the `opf/openproject` repo at `./doc/apiv2-documentation.md`
+- there is no documentation on the experimental API
+
+To get a feel for which API is used at which point, please refer to the `PathHelper` located at `./frontend/app/helpers/path-helper.js`. It is used throughout the application to centralize knowledge about paths.
+
+## HAL
+
+While having a `PathHelper` certainly helps, the long-term idea is to levergae the [HAL](http://stateless.co/hal_specification.html)-capabilities of the APIv3 (thereby excluding `v2` and `experimental` long term) to let any client discover th paths in the api by inspecting the responses from any given call.
```
Comment (jonasheinrich) [inline frontend/doc/API.md]: leverage?

Comment (jonasheinrich) [inline frontend/doc/API.md]: discover **the** paths?

Comment (floriank) [inline frontend/doc/API.md]: yes, and yes, thanks.

**Comment by floriank** [comment]
@basiszwo could you tell me if you have any specific question for the angular frontend.

I would like to add some notes on how the angular frontend can be used in the legacy JS as well as the state of the removal of `prototype.js`

**Comment by basiszwo** [comment]
continue as discussed via skype

**Comment by floriank** [comment]
> continue as discussed via skype

For all the people not privy to that information, I will add:
- notes on Legacy views
- notes on calling our angular services from legacy code
- a note on removing `prototype.js` in the future.

**Comment by basiszwo** [comment]
thank you for the notes.

**Comment by floriank** [comment]
This should conclude documenting the most burning topics for the `angular`based front end. I will proof read this and then it should be merge-able.

**Comment by floriank** [comment]
Proof reading complete.

@opf/developers Please look it over one more time and ask questions - this should be good to merge and will conclude the work for https://community.openproject.org/work_packages/21267.

**Comment by basiszwo** [comment]
I am fine with it.

File: frontend/doc/API.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        "versions": { "href": "/api/v3/projects/1/versions" }
+    },
+    "id": 1,
+    "identifier": "project_identifier",
+    "name": "Project example",
+    "description": "Lorem ipsum dolor sit amet"
+}
+```
+
+The `Project` structure contains links to ressources associated. At the time of writing, there is no ticket endpoint in the API, but just to give an example of using this, given the knowledge about `_links`, one may easily infer the path from the response:
```
Comment (NobodysNightmare) [inline frontend/doc/API.md]: At the time of merging this PR, this information is already otudated xD

We now have an [endpoint to list work packages](http://opf.github.io/apiv3-doc/#work-packages-work-packages-by-project-get).

Comment (floriank) [inline frontend/doc/API.md]: Cool, did not know that - the endpoint is yet to be used. I will shorten the sentence as to not imply that no such things exist.

Double negative ftw.

Comment (basiszwo) [inline frontend/doc/API.md]: good point @NobodysNightmare  :+1:

**Comment by NobodysNightmare** [comment]
I had a quick fly-by, looking nice.

only question for me is the rooting:

Should it be `frontend/doc`? (pro: some day we might want to split frontend/backend into separate repos)
Should it be `doc/frontend`? (pro: that is where everything else lives that is documentation -> discoverability)

**Comment by floriank** [comment]
I cannot decide that, as I just used the original `./frontend/doc` where everything has been living since the dawn of `./frontend` (read: since 4.1)

**Comment by basiszwo** [comment]
Let's keep the files in `frontend/doc`.

**Pull request merged**